cxml/xml.go

```diff
@@ -0,0 +1,222 @@
+/*
+Copyright 2015 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package cxml
+
+import (
+	"encoding/xml"
+	"io"
+	"strings"
+
+	"github.com/gravitational/trace"
+)
+
+// TransofrmFunc is a function that can transform incoming token
+// into a series of outgoing tokens when traversing XML tree
+type TransformFunc func(parents *NodeList, in xml.Token) []xml.Token
+
+// ConditionFunc is accepted by some transform function generators
+// so they could be applied conditionaly
+type ConditionFunc func(parents *NodeList, in xml.Token) bool
+
+// TransformXML parses the XML tree, traverses it and calls TransformFunc
+// on each XML token, writing the output to the writer, resulting in a
+// transformed XML tree
+func TransformXML(r io.Reader, w io.Writer, fn TransformFunc, indent bool) error {
+	parser := xml.NewDecoder(r)
+	encoder := xml.NewEncoder(w)
+	if indent {
+		encoder.Indent("  ", "    ")
+	}
+	parentNodes := &NodeList{}
+	for {
+		token, err := parser.Token()
+		if err != nil {
+			if err != io.EOF {
+				return trace.Wrap(err)
+			}
+			break
+		}
+		for _, t := range fn(parentNodes, token) {
+			if err := encoder.EncodeToken(t); err != nil {
+				return err
+			}
+		}
+		switch e := token.(type) {
+		case xml.StartElement:
+			parentNodes.Push(e)
+		case xml.EndElement:
+			parentNodes.Pop()
+		}
+	}
+	encoder.Flush()
+	return nil
+}
+
+// SetAttribute is XML helper that allows to set attribute on a node
+func SetAttribute(e xml.StartElement, name, value string) xml.StartElement {
+	if len(e.Attr) != 0 {
+		for i := range e.Attr {
+			if e.Attr[i].Name.Local == name {
+				e.Attr[i].Value = value
+				return e
+			}
+		}
+	} else {
```

Comment:
This will miss adding a new attribute on an element with a non-empty attribute list - is this intentional?